From e073271f53583f2d0cf2675ea665ed50712b65dd Mon Sep 17 00:00:00 2001
From: Klaus Wenninger <klaus.wenninger@aon.at>
Date: Mon, 8 Jan 2018 12:07:33 +0100
Subject: [PATCH] Doc: sbd.8.pod: add sections for query-watchdog &
 test-watchdog

---
 man/sbd.8.pod | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/man/sbd.8.pod b/man/sbd.8.pod
index c1bf67a..ffd01c2 100644
--- a/man/sbd.8.pod
+++ b/man/sbd.8.pod
@@ -394,6 +394,30 @@ a fencing message by the cluster.
 
 =back
 
+=head2 query-watchdog
+
+Example usage:
+
+	sbd query-watchdog
+
+Check for available watchdog devices and print some info.
+
+B<Warning>: This command will arm the watchdog during query, and if your
+watchdog refuses disarming (for example, if its kernel module has the
+'nowayout' parameter set) this will reset your system.
+
+=head2 test-watchdog
+
+Example usage:
+
+	sbd test-watchdog [-w /dev/watchdog3]
+
+Test specified watchdog device (/dev/watchdog by default).
+
+B<Warning>: This command will arm the watchdog and have your system reset
+in case your watchdog is working properly! If issued from an interactive
+session, it will prompt for confirmation.
+
 =head1 Base system configuration
 
 =head2 Configure a watchdog
-- 
1.8.3.1

